### ISSUE TITLE
Potentially correlated random number streams

### ISSUE THREAD

**Issue by aaronpeikert** [cjvanlissa/bayesynth#11]
While I was aimlessly scrolling through your code, I noticed that you use a *potentially* unwise method for random number generation. The seeds for each replication are chosen randomly between each possible integer:

https://github.com/cjvanlissa/bayesynth/blob/878da0627b4c21e6358d42169f529cbfef13993d/Sim/Simulatie_V2.R#L35

And are latter used as is:

https://github.com/cjvanlissa/bayesynth/blob/878da0627b4c21e6358d42169f529cbfef13993d/Sim/Simulatie_V2.R#L53

This approach does not create independent random number streams. In fact, it is very likely that the seeds are very far apart, which perhaps unintuitively makes the streams more likely to be correlated. The reason is that R's standard RNG "Mersenne Twister" creates independent streams only when seeds are close together and is generally not suited to create multiple streams used in parallel.
To get around this problem, two things need to be changed:
 1. an RNG capable of creating independent streams must be used.
 2. the seeds for this RNG must be chosen properly.

One way of implementing this without changing much would be:

```r
library(parallel)
RNGkind("L'Ecuyer-CMRG") # 1. RNG capable of creating multiple streams
set.seed(42)
seeds <- purrr::accumulate(
  .x = 1:10,
  .f = function(s, unused)parallel::nextRNGStream(s),  # 2. seeds are ensured to lead to a new stream
  .init = .Random.seed
)
```

**Comment by cjvanlissa** [comment]
Thanks @aaronpeikert ! Useful advice, I'll use this in my next simulations.

**Comment by cjvanlissa** [comment]
Dear Aaron,

I'm looking into this now, and one thing I'm unsure of is the distinction between streams and seeds.
My goal is to make a data.frame with the values of simulation parameters on rows, including a value for the random seed. This allows me to re-run/debug specific rows and use the identical seed.

Yet the code above gives me a list with different vectors of random numbers. How do I plug this into the data.frame?

**Comment by aaronpeikert** [comment]
Hi Caspar,
glad that you pick that up again! Sorry, this is not obvious if you are not familiar with the strange seed system.
if you `set.seed(42)`, `42` is not the actual seed. The acutal seed is:

```r
> set.seed(42)
> .Random.seed
  [1]       10403         624   507561766  1260545903  1362917092 -1772566379
  [7] -1344458670  1072028587 -1293707536  1708789297 -1333339586   256895399
 [13]   998413756  -555191411 -1047547670   -50950813 -1543487672 -1766763351
 [19]    32923734  1971684063  1718493076 -1117926011   890393730 -1032174053
 [25]   932479648 -1892763103 -1040205458   262148887 -1220643732  1566094973
 [31]  -137659622  1045161427 -1555709704   178265753 -1033943674 -1034388913
 [37] -1721830332 -1749741963 -1137867598 -2065544053   866474576   458187281
 [43]  1218274462 -2000447353   199687964  1107998061   590795082  -858760637
 [49]  -240079192   839017609 -1886488906 -1479389761  1595536628  1848478565
 [55]   368165090 -1706758405  -128901632   360384001  2027084750  1372682743
 [61] -1578682932 -1716686755  1324659578  1519053491  2044463192  -906549639
 [67]  1851192294 -1175847633 -1083563612  -849984427   333580562  1871261035
 [73]  1972555184  1756407281  1943215870 -1751074969  1282021500 -1397758131
 [79]   518529450 -1421699293   488820232  -413309847  1749794070   441878687
 [85]    81427028  1971206469 -1203675838   995493851  -366646944  -810597919
 [91]  2006003246  1288196311  -111499476  -280946115    -4973606    75390867
 [97]  1671430072  -422306215 -1145056698  -393285617  1823866628  1253337653
[103]  1712504178  1846705739 -1820069616  -990520879   247466334 -1717615033
[109]  1633935836  -246413011  1455137290 -1652210685   741252456  1555715145
[115]   197538678 -1312148609  -687362124  1153947429   286861730   568692923
[121]  1575576768  1925410241  1087549582  1279039415 -1208956788  1317778461
[127] -1306414022   257836147  1550427928    86967865 -1875779418  -520138001
[133]  1989839972  1598521365  1842848210 -1967767765 -1806646160 -1558774351
[139] -1040440386  1215776039  1529633596 -1637672179  -121023894  -992976669
[145] -1867768632 -1273940951  1167448534   786779743 -2131127020 -1987943163
[151]   563160578  1714129307 -1871376352 -1347450463   590308078  -135901545
[157] -2080269844  1678065149 -1705673574  1784846675  -871301512  1157619225
[163]   491069190   301147599  -563099196 -1769503243  -230115790  1806462987
[169]  1792100304  1913393553    35614238 -1176456185  -165968740   -27190035
[175] -1087819062  1583082947   678644776 -1957034999   760393782   778033471
[181]  -636999052   757457637  -155102622 -1134562693 -1432327296   862688641
[187]   960447822  1400730999 -1343939764 -1742358563  2020046074   549674547
[193] -2039609896  1111401977  -527331994  -970823505  -779241692 -1209238571
[199]  -964822382  1395462379  -170034384 -1664281231   264366206  1603506919
[205] -1702273540   398593741  -269270230 -1017124189  1029450632   -57883671
[211]   647280278   686937119  -443846700  1428836549 -1446924606  1958398811
[217] -1052543264 -1494249119  1771722670 -1095103401   972246188   258286013
[223] -1737515686  1505268499  -883376840   350445017 -1548801082   608508815
[229] -1454615420  -957455947  -913346830   527442379     5070480 -1974335151
[235]  -142896418   122151367  1556997980 -1392643923  1564443530  1762940803
[241] -2059486488 -1650363447  -520885002  1934835455  -757373644  1661447845
[247]  1604991778  2071204923  -717868480 -1355580095  1725471246  -179039433
[253]  -863752692 -1438942307  -762972742  1484404723  1185490072  1357252025
[259]  2083912230   789790319  -352083484    50673557  -426437806  1252893355
[265]  1090056688 -1651315407 -1847300802  -625631065   -46063428  1015857805
[271]  1696986090  -525257629   569571912 -2138041431  1476940630  1044126175
[277]   -45086060  -197788539  1234368386  1489227035  -749691488  -315664095
[283] -1349383058   251890199 -1108361364    86033789 -1966965222  2084588755
[289]   272055288   104766873  -876742522  -985345713  1208721220  -408355467
[295]   346482610  -350383221  1522021712   898089233 -2087422050  1603587975
[301]  -398783972    15107181  -239168438  -680623805 -1588532824   860922761
[307]  -748033610 -1771805505  -331259916  -500352411 -1533811742   820114939
[313] -1804945152   -35969791 -1906397490  2139123959   277741772  2022506333
[319] -1321388422  1078120883 -1411710120 -1058724487  1071589094 -1584278481
[325] -1648603996   743551829  1457318930 -1292372885  -597481296 -1389853455
[331]  1025749502  2006806119   907256700  -259836339  2032231594   377765411
[337]   -47150840 -1076157591  -424628202  1686380447  1392993620  1113942085
[343] -1178271678 -1006203173  -581139360  2049892577   433488174   377669591
[349]  1924592172   418916669 -1039261990  1002030739   158104248 -1999528615
[355]  -768506554  1521633039  -177362428  -998811339 -1135665038  -260782773
[361]  1087491088  1533884625   -81126306  1619492167 -1423774500 -1109731283
[367]   -43621110 -2094372093 -1687562136 -1406692535  1903470198  1834175103
[373]   284826292  1716946469  -566342494  1852413883  1593704384  -148730687
[379]   881951630    -4026697  1052939148 -1227209955  -984795334   568142707
[385] -2067553768  -503577287  -934472794  1667115503 -2072529052  -544084203
[391]  1612022994 -1916008917  -287563920 -1823613775  -968902978 -1274348505
[397] -1212094916  -681219571   212178282   513345507  1305794504   -97651927
[403] -1622291242  1367991647   829521940  -612854779  1930938626   675483803
[409] -1238947040    95299745 -1926777362 -1023949417  2064180460  -559198979
[415]  1326612378 -1041956781  -440895112  -866362087 -1278619130   172850383
[421] -1405979452  -384207627  1766332722   288732939   962208464 -1547538287
[427]  1929150750  1742727943  2017824668  1738206189 -1057557046   101192899
[433]  1380550440   669401865  -405527754 -1914703809  -339372684  1799590373
[439]  -446073502 -2050295429  1806698112   852079745 -1640949682  1138388087
[445]  -639506868  -656193827  2138437626   118048051  1612906712 -1008032519
[451]  1916780646  2050506671  -101326300 -2004489515   184475026 -1662396437
[457]  1596184112  -675089295 -1577550978  -843167257 -1257707268  1425235405
[463] -1066236374  1924108707  1386210952   793281257   245344662  2064476959
[469] -1555146028   456097733 -1370795582 -1220980133  -193949216   124596321
[475] -1371166034 -1037925545 -1180330068 -1443221315    42967130  -125699565
[481] -1814349768 -1063330599   759619270 -1177128305   656015236 -1588637515
[487] -2075012622  -283088693 -1961805424  1984391249 -1077172770 -1822549817
[493]  -496831908  1105641389  1026573962 -1178108285  1689253352 -2041793847
[499]   591945718  1305105919  -412889036   750017957  1389714978 -2000282821
[505] -1321153216   143694913  -805475058  -645395913   615250188   308808349
[511]   276265146 -1182327053 -1934024808   913376441  1517759782 -1411377809
[517]   418827492  1391306389   542701138  1645308331  -738571024 -1035482063
[523]    84803646 -1032366169   548121532 -1930620531  -179816726  1283971939
[529]   173126984   518705833 -2124003754   382644447  1935537556   491400069
[535]  1679792770  1655264283  -285689696 -1221855199  -504340626 -2117926633
[541]  -783480212 -1901800323  1833271578 -1991199789 -1030441224   358161561
[547] -1150768250   252520527  -521909692  -120001419   908872370  -336274805
[553]  1018630224   -88334319  1985448606 -1061026169  1010504988  1350456173
[559]   852645706 -1205295037   828187816  1729815177  -795779914 -1026393153
[565]   781503220 -1503073947  2135033570  1226505467  -428846080 -1875426303
[571] -1900641842   244017143   568380364  1462275677  1762769274 -1021921101
[577]   424017496  -917560199  1572035046  2115349295 -1137018972   713630293
[583]   686018322   520272747 -1273003088  1520198641  -465550082  1341531495
[589] -1385615748  1662158157  -874076246 -1511922397   866796552  1221911145
[595]    73507606   445494943   744509524 -1115121851  1297392450  -598534693
[601] -1132486816   212500449  1290261550   798572247   616512812  1617639485
[607]  -537648678  -569299565  -526060104   978000985 -1695598522  1873911311
[613]   340874500 -1149876171  1737982834   695405643   353085200   417372113
[619]  -346017954 -1916029881 -1735526436  1461823277   580724746  -622095869
[625]  -686746776   705745481
> 
```

An integer vector! Like you get as elements in the list `seeds` above. Simply use `set.seed` with the integer vector, i.e., `set.seed(seeds[[2]])`.

```
library(parallel)
library(tibble)

RNGkind("L'Ecuyer-CMRG") # 1. RNG capable of creating multiple streams
set.seed(42)
seeds <- purrr::accumulate(
  .x = 1:10,
  .f = function(s, unused)parallel::nextRNGStream(s),  # 2. seeds are ensured to lead to a new stream
  .init = .Random.seed
)

simulation = tibble::tibble(seed = seeds, simulation_result = purrr::map(seed, function(seed){set.seed(seed);mean(rnorm(30))}))
```

**Comment by cjvanlissa** [comment]
Sorry but I'm still not entirely sure how I get a vector of, e.g. 100000 random numbers this way... I'm having trouble unpacking the purr::accumulate statement. I'd like the seeds to be part of the data.frame of simulation conditions, so I can reproduce individual runs if necessary...

**Comment by aaronpeikert** [comment]
Apparently this does not work :cry: 

Replace the last line with:

`simulation = tibble::tibble(seed = seeds, simulation_result = purrr::map_dbl(seed, function(seed)with(list(.Random.seed = seed), mean(rnorm(30)))))`

**Comment by aaronpeikert** [comment]
> Sorry but I'm still not entirely sure how I get a vector of, e.g. 100000 random numbers this way... I'm having trouble unpacking the purr::accumulate statement. I'd like the seeds to be part of the data.frame of simulation conditions, so I can reproduce individual runs if necessary...

the accumulate statement is replacing this step: `sample(1:.Machine$integer.max, nrow(summarydata))`.
Instead of using random numbers to sample seeds, you use the well-defined mechanism of nextRNGStream.
 Which returns the state of the RNG to which you can return. So you can reproduce individual runs by resetting the state of the RNG by e.g. to reproduce the third simulation `.Random.seed <- simulation$seed[[3]]`

**Comment by cjvanlissa** [comment]
That part I understand... but I'm struggling with the tibble and purrr statements.

**Comment by aaronpeikert** [comment]
Ok I'll replace the tidyverse style stuff with base R :smiling_imp:

**Comment by cjvanlissa** [comment]
Sorry to be a bore :P It might be the heat, but I'm really struggling to grok

**Comment by aaronpeikert** [comment]
I hope this is somewhat clearer?
```
RNGkind("L'Ecuyer-CMRG") # 1. RNG capable of creating multiple streams
set.seed(42)
seeds <- vector("list", 10)
seeds[[1]] <- .Random.seed
for (i in 2:length(seeds)) {
  seeds[[i]] <- nextRNGStream(seeds[[i - 1]])
}
# grid is created after seeds;
# that way, corner cases of the random process are replicated across conditions,
# only relevant in simulations with e.g. random walks
hyper_parameters <- list(es = seq(0, 1, .1), n = 100, seed = seeds)
summarydata <- expand.grid(hyper_parameters, stringsAsFactors = FALSE)

for (i in 1:nrow(summarydata)) {
  attach(summarydata[i, ]) # this should be replaced by with(summarydata[i, ], {code})
  .Random.seed <- seed #this is where the magic happens
  t.test(x = rnorm(n), y = rnorm(n) + es)$p.value
}
```

To reproduce a condition:

```
attach(summarydata[3, ])
.Random.seed <- seed
```

**Comment by cjvanlissa** [comment]
I'm still wondering: why is the seed now included in expand.grid? The sim has 11 conditions, but each condition is now repeated 10 times (because 10 rng states were generated).

What I would expect is:

The sim has 11 conditions, generate 11 rng states, run each once?!

**Comment by aaronpeikert** [comment]
Yes, that's right and intentional. Note that this has nothing to do with the correlated random number streams but is a separate issue. You can exclude the seeds from the `expand.grid` without any problem if you want to. The problem of correlated streams is solved.

```
# grid is created after seeds;
# that way, corner cases of the random process are replicated across conditions,
# only relevant in simulations with e.g. random walks
```

I think it is generally better to generate a seed for each repetition instead of each condition. The reasoning is that, as with an experiment in the real world, you want to minimize differences between conditions that are not a direct result of the manipulation. Imagine you have conditions A and B, with their own seed. Then you might run into the situation that in condition A, everything is fine, every model converges, etc., but in condition B, by pure chance, you generate data that lets a model fail to converge. You could even blame condition B for it (though it is random), or even worse, you might calculate the parameter bias as the average deviation, and since you have a huge outlier due to non-convergence, you conclude that B is not a good fit for these types of models. Of course, these types of errors average out over time with increased sample size, but there is no need to run the risk, is there?
The only downside, I think, is very theoretical. Such setup implies that any hypothesis test should account for the dependence across seeds, so instead of doing a t.test on the results, you'd do a paired t.test. This setup is a special case of a multilevel model where the design is fully crossed, and each condition was observed with each seed. Therefore, as long as ergodicity holds, the t.test just has less power but holds the alpha, and I think it is quite hard to argue that ergodicity doesn't hold.
